astropy/coordinates/solar_system.py

```diff
@@ -56,116 +68,91 @@ def _download_spk_file(url=('http://naif.jpl.nasa.gov/pub/naif/'
     global KERNEL
 
-    try:
-        from jplephem.spk import SPK
-    except ImportError:
-        raise ImportError("Solar system ephemeris calculations require the "
-                          "jplephem package "
-                          "(https://pypi.python.org/pypi/jplephem)")
+    ephemeris = ephemeris.lower()
+
+    if KERNEL is None or KERNEL.name != ephemeris:
+        try:
+            from jplephem.spk import SPK
+        except ImportError:
+            raise ImportError("Solar system JPL ephemeris calculations require "
+                              "the jplephem package "
+                              "(https://pypi.python.org/pypi/jplephem)")
 
-    if KERNEL is None:
-        KERNEL = SPK.open(_download_spk_file())
+        KERNEL = SPK.open(_download_spk_file(ephemeris))
+        KERNEL.name = ephemeris
     return KERNEL
 
 
-def get_body_barycentric(time, body):
+def get_body_barycentric(body, time, ephemeris=None):
     """
     Calculate the barycentric position of the solar system body ``body``
     in cartesian coordinates.
 
-    Uses ``jplephem`` with the DE430 kernel.
-
     Parameters
     ----------
-    time : `~astropy.time.Time`
-        Time of observation
-
     body : str
-        The solar system body to calculate.
-
-        The allowed values for ``body`` can be found in
+        The solar system body to calculate. The allowed values for ``body``
+        depend on the ephemeris, but standard ones can be found in
         ``astropy.coordinates.SOLAR_SYSTEM_BODIES``.
+    time : `~astropy.time.Time`
+        Time at which the position should be calculated.
 
     Returns
     -------
     cartesian_position : `~astropy.coordinates.CartesianRepresentation`
         Barycentric (ICRS) position of the body in cartesian coordinates
 
-    Notes
-    -----
-
     """
-    # lookup chain
-    chain = BODY_NAME_TO_KERNEL_SPEC[body.lower()]
-
-    kernel = _get_kernel()
-
     jd1, jd2 = get_jd12(time, 'tdb')
+    body = body.lower()
+    if ephemeris is None:
+        earth_pv_helio, earth_pv_bary = erfa.epv00(jd1, jd2)
+        if body == 'earth':
+            cartesian_position_body = earth_pv_bary[..., 0, :]
+
+        else:
+            sun_bary = earth_pv_bary[..., 0, :] - earth_pv_helio[..., 0, :]
+            if body == 'sun':
+                cartesian_position_body = sun_bary
+            else:
+                body_index = ERFA_BODY_NAME_TO_PLANET_INDEX[body]
+                body_pv_helio = erfa.plan94(jd1, jd2, body_index)
+                cartesian_position_body = body_pv_helio[..., 0, :] + sun_bary
+
+        barycen_to_body_vector = u.Quantity(
+            np.rollaxis(cartesian_position_body, -1, 0), u.au)
 
-    cartesian_position_body = sum([kernel[pair].compute(jd1, jd2) for pair in chain])
-
-    barycen_to_body_vector = u.Quantity(cartesian_position_body, unit=u.km)
-    return CartesianRepresentation(barycen_to_body_vector)
-
-
-def _get_earth_body_vector(time, body, earth_time=None):
```

Comment:
For clarification, I combined this in the code below to avoid recalculating the position of the earth many times.